stl/inc/ranges

```diff
@@ -4748,6 +4748,378 @@ namespace ranges {
+            if constexpr (_Slide_caches_first<_Vw>) {
+                auto _First = _RANGES begin(_Range);
+                if (this->_Has_cache()) {
+                    return _Iterator<false>{_First, this->_Get_cache(_Range), _Count};
+                }
+
+                auto _Result = _RANGES next(_First, _Count - 1, _RANGES end(_Range));
+                this->_Set_cache(_Range, _Result);
+                return _Iterator<false>{_First, _Result, _Count};
+            } else {
+                return _Iterator<false>{_RANGES begin(_Range), _Count};
+            }
+        }
+
+        _NODISCARD constexpr auto begin() const noexcept(
+            noexcept(_RANGES begin(_Range)) && is_nothrow_move_constructible_v<iterator_t<_Vw>>) /* strengthened */
+            requires _Slide_caches_nothing<const _Vw> {
+            return _Iterator<true>{_RANGES begin(_Range), _Count};
+        }
+
+        // clang-format off
+        _NODISCARD constexpr auto end() requires (!(_Simple_view<_Vw> && _Slide_caches_nothing<const _Vw>) ) {
+            // clang-format on
+            if constexpr (_Slide_caches_nothing<_Vw>) {
+                const auto _Size = _RANGES distance(_Range) - (_Count - 1);
+                if (_Size < 0) {
+                    return _Iterator<false>{_RANGES begin(_Range), _Count};
+                }
+
+                return _Iterator<false>{_RANGES begin(_Range) + _Size, _Count};
+            } else if constexpr (_Slide_caches_last<_Vw>) {
+                if (this->_Has_cache()) {
+                    return _Iterator<false>{this->_Get_cache(_Range), _Count};
+                }
+
+                auto _Result = _RANGES prev(_RANGES end(_Range), _Count - 1, _RANGES begin(_Range));
+                this->_Set_cache(_Range, _Result);
+                return _Iterator<false>{_Result, _Count};
+            } else if constexpr (common_range<_Vw>) {
+                const auto _Last = _RANGES end(_Range);
+                return _Iterator<false>{_Last, _Last, _Count};
+            } else {
+                return _Sentinel{_RANGES end(_Range)};
+            }
+        }
+
+        _NODISCARD constexpr auto end() const noexcept(is_nothrow_move_constructible_v<iterator_t<_Vw>> //
+                && noexcept(_RANGES distance(_Range)) && noexcept(_RANGES begin(_Range) + _Count)) /* strengthened */
+            requires _Slide_caches_nothing<const _Vw> {
+            const auto _Size = _RANGES distance(_Range) - (_Count - 1);
+            if (_Size < 0) {
+                return _Iterator<true>{_RANGES begin(_Range), _Count};
+            }
+
+            return _Iterator<true>{_RANGES begin(_Range) + _Size, _Count};
+        }
+
+        _NODISCARD constexpr auto size() noexcept(noexcept(_RANGES distance(_Range))) /* strengthened */
+            requires sized_range<_Vw> {
+            const auto _Size = _RANGES distance(_Range) - (_Count - 1);
+            if (_Size < 0) {
+                return _To_unsigned_like(_Size - _Size);

```

Comment:
I think I would prefer the standard way of doing this; this is a little confusing and I hate using `x - x` to get `0` of the right type.